frontend/lib/src/dataframes/arrowFormatUtils.ts

```diff
@@ -144,89 +151,147 @@ interface Interval {
-    return Number(value / BigInt(unitAdjustment))
+  if (
+    typeof timestamp === "bigint" &&
+    !Number.isSafeInteger(Number(timestamp))
+  ) {
+    return Number(timestamp / BigInt(unitAdjustment))
   }
 
-  return Number(value) / unitAdjustment
+  return Number(timestamp) / unitAdjustment
 }
 
-function formatTime(data: number, field?: Field): string {
-  const timeInSeconds = convertToSeconds(data, field?.type?.unit ?? 0)
-  return moment
-    .unix(timeInSeconds)
+/**
+ * Converts a timestamp to a date object.
+ *
+ * @param timestamp The timestamp to convert.
+ * @param field The field containing the unit information.
+ * @returns The date object.
+ */
+export function convertTimestampToDate(
+  timestamp: number | bigint,
+  field?: Field
+): Date {
+  // Time values from arrow are not converted to a shared unit and
+  // just return the raw arrow value. Therefore, we need to adjust
+  // the value to seconds based on the unit information in the field.
+  // https://github.com/apache/arrow/blob/9e08c57c0986531879aadf7942998d26a94a5d1b/js/src/visitor/get.ts#L193C7-L209
+  const timeInSeconds = convertTimestampToSeconds(
+    timestamp,
+    field?.type?.unit ?? TimeUnit.SECOND
+  )
+  return moment.unix(timeInSeconds).utc().toDate()
+}
+
+/**
+ * Formats a time value based on the unit information in the field.
+ *
+ * @param timestamp The time value to format.
+ * @param field The field containing the unit information.
+ * @returns The formatted time value.
+ */
+function formatTime(timestamp: number | bigint, field?: Field): string {
+  const date = convertTimestampToDate(timestamp, field)
+  return moment(date)
     .utc()
-    .format(timeInSeconds % 1 === 0 ? "HH:mm:ss" : "HH:mm:ss.SSS")
+    .format(date.getMilliseconds() === 0 ? "HH:mm:ss" : "HH:mm:ss.SSS")
 }
 
-function formatDuration(data: number | bigint, field?: Field): string {
+function formatDate(date: number | Date): string {
+  const formatPattern = "YYYY-MM-DD"
+
+  // Date values from arrow are already converted to a date object
+  // or a timestamp in milliseconds even if the field unit might indicate a
+  // different unit.
+  // https://github.com/apache/arrow/blob/9e08c57c0986531879aadf7942998d26a94a5d1b/js/src/visitor/get.ts#L167-L171
+  if (
+    date instanceof Date ||
+    (typeof date === "number" && Number.isFinite(date))
+  ) {
+    return moment.utc(date).format(formatPattern)
+  }
+
+  logWarning(`Unsupported date type: ${date}`)
+  return String(date)
+}
+
+/**
+ * Format datetime value from Arrow to string.
+ */
+function formatDatetime(date: number | Date, field?: Field): string {
+  // Datetime values from arrow are already converted to a date object
+  // or a timestamp in milliseconds even if the field unit might indicate a
+  // different unit.
+  // https://github.com/apache/arrow/blob/9e08c57c0986531879aadf7942998d26a94a5d1b/js/src/visitor/get.ts#L174-L190
+  let datetime = moment.utc(date)
```

Comment:
Why don't we have to do the same 

```typescript
    date instanceof Date ||
    (typeof date === "number" && Number.isFinite(date))
```

check we do in `formatDate` here?

Reply:
Good point, I also added this check to datetime as well.

Reply:
great thanks 🙂 might be worth to have a dedicated function for this check in this file but no strong opinion